lib/mayaUsd/ufe/UsdUndoRenameCommand.cpp

```diff
@@ -84,95 +105,83 @@ UsdSceneItem::Ptr UsdUndoRenameCommand::renamedItem() const
+        auto newUfePath = Ufe::Path(newSceneItem->path().getSegments()[1]);
+        bool status = MayaUsdUtils::updateInternalReferencesPath(prim, SdfPath(newUfePath.string()));
+        if (!status) {
+            return false;
+        }
 
-    // set prim's name
-    // XXX: SetName successfuly returns true but when you examine the _prim.GetName()
-    // after the rename, the prim name shows the original name HS, 6-May-2020.
-    bool status = primSpec->SetName(_newName);
-    if (!status) {
-        return false;
+        // set the new name
+        auto primSpec = MayaUsdUtils::getPrimSpecAtEditTarget(prim);
+        status = primSpec->SetName(_newName);
+        if (!status) {
+            return false;
+        }
     }
 
     // the renamed scene item is a "sibling" of its original name.
     _ufeDstItem = createSiblingSceneItem(_ufeSrcItem->path(), _newName);
 
-    sendNotification<Ufe::ObjectRename>(_ufeDstItem, _ufeSrcItem->path());
-
-    // SdfLayer is a "simple" container, and all it knows about defaultPrim is that it is a piece of token-valued layer metadata.  
-    // It is only the higher-level Usd and Pcp modules that know that it is identifying a prim on the stage.  
-    // One must use the SdfLayer API for setting the defaultPrim when you rename the prim it identifies.
-    if(primPath == defaultPrimPath){
+    // update stage's default prim
+    if(_ufeSrcItem->prim().GetPath() == defaultPrimPath) {
         _stage->SetDefaultPrim(_ufeDstItem->prim());
     }
 
+    // send notification to update UFE data model
+    sendNotification<Ufe::ObjectRename>(_ufeDstItem, _ufeSrcItem->path());
+
     return true;
 }
 
 bool UsdUndoRenameCommand::renameUndo()
 {
-    const UsdPrim& prim = _stage->GetPrimAtPath(_ufeDstItem->prim().GetPath());
+    // get the stage's default prim path
+    auto defaultPrimPath = _stage->GetDefaultPrim().GetPath();
 
-    auto primSpec = MayaUsdUtils::getPrimSpecAtEditTarget(prim);
-    if(!primSpec) {
-        return false;
-    }
+    // 1- open a changeblock to delay sending notifications.
+    // 2- update the Internal References paths (if any) first
+    // 3- set the new name 
+    {
+        SdfChangeBlock changeBlock;
 
-    // these two lines MUST be called before the set name
-    // _stage->GetDefaultPrim() and prim after the rename can be invalid.
-    auto primPath = prim.GetPath();
-    auto defaultPrimPath = _stage->GetDefaultPrim().GetPath();
+        const UsdPrim& prim = _stage->GetPrimAtPath(_ufeDstItem->prim().GetPath());
 
-    // set prim's name
-    bool status = primSpec->SetName(_ufeSrcItem->prim().GetName());
-    if (!status) {
-        return false;
+        auto newSceneItem = createSiblingSceneItem(_ufeSrcItem->path(), _ufeSrcItem->prim().GetName());
+        auto newUfePath = Ufe::Path(newSceneItem->path().getSegments()[1]);
+        bool status = MayaUsdUtils::updateInternalReferencesPath(prim, SdfPath(newUfePath.string()));
```

Comment:
Same comment as redo().  In fact, undo() and redo() are so similar it would be good to factor out the code into a separate private method and call that from undo() and redo() with the proper arguments.